File: app/src/main/java/com/wegielek/signalychinese/views/HistoryActivity.java
package com.wegielek.signalychinese.views;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;

import android.content.Intent;
import android.os.Bundle;
import android.view.MenuItem;
import android.widget.Toast;

import com.wegielek.signalychinese.R;
import com.wegielek.signalychinese.adapters.HistoryAdapter;
import com.wegielek.signalychinese.database.SearchHistory;
import com.wegielek.signalychinese.databinding.ActivityHistoryBinding;
import com.wegielek.signalychinese.interfaces.HistoryRecyclerViewListener;
import com.wegielek.signalychinese.viewmodels.HistoryViewModel;

import java.util.ArrayList;
import java.util.List;

public class HistoryActivity extends AppCompatActivity implements HistoryRecyclerViewListener {

    private ActivityHistoryBinding mBinding;
    private HistoryAdapter mHistoryAdapter;
    private HistoryViewModel mHistoryViewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mBinding = DataBindingUtil.setContentView(this, R.layout.activity_history);

        mHistoryViewModel = new ViewModelProvider(this).get(HistoryViewModel.class);

        Toolbar historyToolbar = mBinding.historyToolbar;
        setSupportActionBar(historyToolbar);
        historyToolbar.setTitleTextColor(getColor(R.color.dark_mode_white));
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            getSupportActionBar().setDisplayShowHomeEnabled(true);
            getSupportActionBar().setTitle("History");
        }

        mBinding.historyRv.setLayoutManager(new LinearLayoutManager(this));
        mHistoryAdapter = new HistoryAdapter(this, this);
        mBinding.historyRv.setAdapter(mHistoryAdapter);


        mHistoryViewModel.getWholeHistory().observe(this, searchHistories -> {
            List<String> searchHistoryList = new ArrayList<>();
            for (SearchHistory searchHistory : searchHistories) {
                searchHistoryList.add(searchHistory.time + "/" + searchHistory.traditionalSign + "/" + searchHistory.simplifiedSign
                        + "/" + searchHistory.pronunciation + "/" + searchHistory.pronunciationPhonetic
                        + "/" + searchHistory.translation);
            }
            mHistoryAdapter.setData(searchHistoryList);
        });

    }

    @Override
    public void onHistoryClicked(String word) {
        Intent intent = new Intent(getBaseContext(), DefinitionWordActivity.class);
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP);
        intent.putExtra("word", word);
        startActivity(intent);
    }

    @Override
    public void onLongHistoryClicked(String timestamp) {
        Toast.makeText(this, timestamp, Toast.LENGTH_SHORT).show();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            finish();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/com/wegielek/signalychinese/views/HistoryActivity.kt
package com.wegielek.signalychinese.views

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.content.ContextCompat
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.wegielek.signalychinese.R
import com.wegielek.signalychinese.adapters.HistoryAdapter
import com.wegielek.signalychinese.database.Dictionary
import com.wegielek.signalychinese.database.History
import com.wegielek.signalychinese.databinding.ActivityHistoryBinding
import com.wegielek.signalychinese.interfaces.HistoryRecyclerViewListener
import com.wegielek.signalychinese.viewmodels.HistoryViewModel

class HistoryActivity : AppCompatActivity(), HistoryRecyclerViewListener {

    private lateinit var mBinding: ActivityHistoryBinding
    private lateinit var mHistoryAdapter: HistoryAdapter
    private lateinit var mHistoryViewModel: HistoryViewModel
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mBinding = ActivityHistoryBinding.inflate(layoutInflater)
        setContentView(mBinding.root)
        mHistoryViewModel = ViewModelProvider(this)[HistoryViewModel::class.java]
        val historyToolbar: Toolbar = mBinding.historyToolbar
        setSupportActionBar(historyToolbar)
        historyToolbar.setTitleTextColor(getColor(R.color.dark_mode_white))
        if (supportActionBar != null) {
            supportActionBar!!.setDisplayHomeAsUpEnabled(true)
            supportActionBar!!.setDisplayShowHomeEnabled(true)
            supportActionBar!!.title = getString(R.string.history)
        }
        mBinding.historyRv.layoutManager = LinearLayoutManager(this)
        mHistoryAdapter = HistoryAdapter(this, this)
        mBinding.historyRv.adapter = mHistoryAdapter
        mHistoryViewModel.wholeHistory.observe(
            this
        ) { histories: List<History> ->
            mHistoryAdapter.setData(
                histories
            )
        }
    }

    override fun onHistoryClicked(dictionary: Dictionary) {
        val intent = Intent(baseContext, DefinitionWordActivity::class.java)
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP)
        intent.putExtra("word", dictionary)
        startActivity(intent)
    }

    override fun onLongHistoryClicked(timestamp: String) {
        Toast.makeText(this, timestamp, Toast.LENGTH_SHORT).show()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.history_menu, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            onBackPressedDispatcher.onBackPressed()
            return true
        } else if (item.itemId == R.id.trash) {
            mHistoryViewModel.deleteWholeHistory().addListener(
                { Log.d(LOG_TAG, "History deleted successfully") },
                ContextCompat.getMainExecutor(this)
            )
        }
        return super.onOptionsItemSelected(item)
    }

    companion object {
        private const val LOG_TAG = "HistoryActivity"
    }
}